### PR TITLE
Issue/postgresql

### PR DESCRIPTION
**Pull request by arnaudsjs**

### CODE REVIEW

File: src/inmanta/data_pg.py [hunk 1]
```diff
@@ -0,0 +1,2165 @@
...
+
+        fields = cls._fields.copy()
+        for name, value in kwargs.items():
+            if name not in fields:
+                raise AttributeError("%s field is not defined for this document %s" % (name, cls.table_name()))
+
+            if value is None and fields[name].required:
+                raise TypeError("%s field is required" % name)
+
+            if not fields[name].reference and value is not None and not (value.__class__ is fields[name].field_type
```
Comment (wouterdb) [inline src/inmanta/data_pg.py]: perhaps consider using https://magicstack.github.io/asyncpg/current/usage.html#asyncpg-custom-codecs

File: src/inmanta/data_pg.py [hunk 2]
```diff
@@ -0,0 +1,2165 @@
...
+                elif from_postgres and isinstance(value, str) and issubclass(fields[name].field_type, enum.Enum):
+                    value = fields[name].field_type[value]
+                else:
+                    raise TypeError("Field %s should have the correct type (%s instead of %s)" %
+                                    (name, fields[name].field_type.__name__, type(value).__name__))
+
+            if value is not None:
+                result[name] = value
+
+            elif fields[name].default:
```
Comment (wouterdb) [inline src/inmanta/data_pg.py]: default here or in schema?

File: src/inmanta/data_pg.py [hunk 3]
```diff
@@ -0,0 +1,2165 @@
...
+
+    def _get_field(self, name):
+        if hasattr(self.__class__, name):
+            field = getattr(self.__class__, name)
+            if isinstance(field, Field):
+                return field
+
+        return None
+
+    def __getattribute__(self, name):
```
Comment (wouterdb) [inline src/inmanta/data_pg.py]: why is this needed?

File: src/inmanta/data_pg.py [hunk 4]
```diff
@@ -0,0 +1,2165 @@
...
+
+    @classmethod
+    async def _execute_query(cls, query, *values, connection=None):
+        if connection:
+            return await connection.execute(query, *values)
+        async with cls._connection_pool.acquire() as con:
+            return await con.execute(query, *values)
+
+    @classmethod
+    async def insert_many(cls, documents):
```
Comment (wouterdb) [inline src/inmanta/data_pg.py]: perhaps use 

https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.executemany

or 

https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.copy_records_to_table

File: src/inmanta/db/versions/v1.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
+async def update(connection):
+    schema = """
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+
+DROP TABLE IF EXISTS project CASCADE;
```
Comment (wouterdb) [inline src/inmanta/db/versions/v1.py]: isn't this very risky to put in the update script that should only every run if the DB is empty?

**Comment by wouterdb** [comment]
we should also document the postgres migration procedure, including
1. install postgress
2.  create db and user 

```
sudo -u postgress -i
 createuser --pwprompt inmanta
 createdb -O inmanta inmanta
```

3. modify pg_hba.conf to set identification to md5 for localhost
4. install uuid-ossp

```
CREATE EXTENSION "uuid-ossp";
```

File: src/inmanta/data_pg.py (hunk 4 above)
Comment (arnaudsjs) [reply src/inmanta/data_pg.py]: New implementation uses copy_records_to_table.

File: src/inmanta/db/versions/v1.py (hunk 5 above)
Comment (arnaudsjs) [reply src/inmanta/db/versions/v1.py]: Drop statements are removed.

**Comment by wouterdb** [comment]
- [ ] backport master to PG
- [ ] process comments on schema
- [ ] is bulk insert all or nothing
- [ ] how do triggers interact with copy to table
- [ ] check on DB to find missing indices
